ros/src/waypoint_updater/waypoint_updater.py

```diff
@@ -62,16 +72,51 @@ def get_waypoint_velocity(self, waypoint):
     def set_waypoint_velocity(self, waypoints, waypoint, velocity):
         waypoints[waypoint].twist.twist.linear.x = velocity
 
-    def distance(self, waypoints, wp1, wp2):
-        dist = 0
-        dl = lambda a, b: math.sqrt((a.x-b.x)**2 + (a.y-b.y)**2  + (a.z-b.z)**2)
-        for i in range(wp1, wp2+1):
-            dist += dl(waypoints[wp1].pose.pose.position, waypoints[i].pose.pose.position)
-            wp1 = i
-        return dist
+    def distance_to_previous(self, position):
+        return euclidean(get_position(self.previous_pos), get_position(position))
+
+    def waypoint_is_ahead(self, waypoint):
+        w_x, w_y, _ = get_position(waypoint.pose.pose)
+        p_x, p_y, _ = get_position(self.previous_pos)
+        heading = math.atan2(w_y-p_y, w_x-p_x)
+        yaw = euler_from_quaternion(get_orientation(self.previous_pos))[2]
+        angle = yaw-heading
+        return True if math.fabs(angle) < math.pi/4. else False
+
+    def update_next_waypoint(self):
+        it = 0
+        nb_waypoints = len(self.static_waypoints.waypoints)
+        while not self.waypoint_is_ahead(self.static_waypoints.waypoints[self.next_waypoint % nb_waypoints]) and \
+                it < UPDATE_MAX_ITER:
+            self.next_waypoint += 1  # We look at the next one
+
+        # Searching the next waypoint in the full path takes much longer, we want to avoid it as much as possible
+        if it == UPDATE_MAX_ITER:
+            self.search_next_waypoint()
+
+    def search_next_waypoint(self):
+        self.next_waypoint = 0
+        # TODO: Rewrite this in a clearer way
```

Comment:
Pull Requests should avoid `TODOs`. A good practice is either do it now or leave it, but create an issue in GitHub and have a reference to it here as well. So the `TODO` is not forgotten.

Reply:
You can think of @agotterba suggestion to use https://docs.scipy.org/doc/scipy-0.14.0/reference/generated/scipy.spatial.KDTree.html . This could be a separate task.

Reply:
Regarding scipy.spatial.KDTree, One issue that waypoint_updater has is that the point it's looking for should be in front of the car; tl_detector only has to look for the nearest point.  KDTree.query lets you ask for the nearest several points by setting the k argument greater than 1.  You can ask for a few points (4? 12?) to make sure at least one is in front of the car, and iterate through them until you find one that's in front (the points are returned sorted from nearest to farthest).

Reply:
Exactly, this is why I haven't used KDTree eventually.